lms/static/scripts/frontend_apps/utils/youtube.ts

```diff
@@ -0,0 +1,86 @@
+import { useEffect, useState } from 'preact/hooks';
+
+import type { ThumbnailData } from '../components/URLFormWithPreview';
+import { InvalidArgumentError } from '../errors';
+
+/**
+ * Tries to resolve the video ID from a YouTube URL
+ * @return Undefined if the ID could not be resolved
+ */
+export function videoIdFromYouTubeUrl(youTubeUrl: string): string | undefined {
+  // This regexp tries to match the video ID in the next possible URLs
+  //  * youtu.be/{id}
+  //  * {domain}/watch?v={id}[...]
+  //  * {domain}/embed/{id}[...]
+  //  * {domain}/shorts/{id}[...]
+  //  * {domain}/live/{id}[...]
+  // See https://stackoverflow.com/a/9102270 for details
+  const match = youTubeUrl.match(
+    /^.*(youtu\.be\/|v\/|u\/\w\/|embed\/|shorts\/|live\/|watch\?v=|&v=)([^#&?]*).*/
+  );
+
+  return match?.[2];
+}
+
+/**
+ * Tries to match provided URL against known YouTube video URL formats,
+ * throwing an error in case of invalid YouTube URL
+ */
+export function validateYouTubeVideoUrl(youTubeUrl: string): string {
+  let url;
+  try {
+    url = new URL(youTubeUrl);
+  } catch (e) {
+    throw new InvalidArgumentError(
```

Comment:
What would be your preference here?

The exception approach has the benefit that it allows to handle other possible runtime errors transparently, falling back to a default error.

It also allows to have a consistent return type, and handle errors with a "different" approach (by catching them, in this case).

On the other hand, TS mitigates the return type inconsistency, by allowing to return different "shapes" and ensuring they are properly handled by callers.

Also, most validation libraries do not throw errors, and considering we already use that approach in other places, it would be more consistent to do the same here.

Reply:
> What would be your preference here?

My preference would be to indicate the error in the return type, as being the most consistent with other validation functions. See `articleIdFromUserInput`, `validateGrade`, `extractBookId`. These either return `null` for invalid inputs, if they don't distinguish between different reasons, or an error message and some validity indicator field otherwise.

> The exception approach has the benefit that it allows to handle other possible runtime errors transparently, falling back to a default error.

This is true. The flip side is that unexpected errors (ie. ones the programmer did not expect) could get masked, instead of being detected as errors and reported to us (via Sentry, if we had that set up).

In general either approach to handling validation errors would be viable. I think the most important thing is to try to be consistent within the project. If there was a de-facto choice in the ecosystem I'd most likely use that in a new project. I'm not sure there is.

Reply:
> The flip side is that unexpected errors (ie. ones the programmer did not expect) could get masked, instead of being detected as errors and reported to us (via Sentry, if we had that set up).

This is very true.